cluster/src/main/java/org/apache/iotdb/cluster/log/manage/serializable/SyncLogDequeSerializer.java

```diff
@@ -0,0 +1,438 @@
+    serializeMeta(meta);
+  }
+
+  @Override
+  public void append(List<Log> logs, LogManagerMeta meta) {
+    int bufferSize = 0;
+    List<ByteBuffer> bufferList = new ArrayList<>(logs.size());
+    for (Log log : logs) {
+      ByteBuffer data = log.serialize();
+      int size = data.capacity() + Integer.BYTES;
+      logSizeDeque.addLast(size);
+      bufferSize += size;
+
+      bufferList.add(data);
+    }
+
+    ByteBuffer finalBuffer = ByteBuffer.allocate(bufferSize);
+    for (ByteBuffer byteBuffer : bufferList) {
+      finalBuffer.putInt(byteBuffer.capacity());
+      finalBuffer.put(byteBuffer.array());
+    }
+
+    // write into disk
+    try {
+      ReadWriteIOUtils.writeWithoutSize(finalBuffer, currentLogOutputStream);
+    } catch (IOException e) {
+      logger.error("Error in log serialization: " + e.getMessage());
+    }
+
+    serializeMeta(meta);
+  }
+
+  @Override
+  public void removeLast(LogManagerMeta meta) {
+    truncateLogIntern(1);
+    serializeMeta(meta);
+  }
+
+  @Override
+  public void truncateLog(int count, LogManagerMeta meta) {
+    truncateLogIntern(count);
+    serializeMeta(meta);
+  }
+
+  private File getCurrentLogFile() {
+    return logFileList.get(logFileList.size() - 1);
+  }
+
+  private void truncateLogIntern(int count) {
+    if (logSizeDeque.size() < count) {
+      throw new IllegalArgumentException("truncate log count is bigger than total log count");
+    }
+
+    int size = 0;
+    for (int i = 0; i < count; i++) {
+      size += logSizeDeque.removeLast();
+    }
+    // truncate file
+    while (size > 0) {
+      File currentLogFile = getCurrentLogFile();
+      // if the last file is smaller than truncate size, we can delete it directly
+      if (currentLogFile.length() < size) {
+        size -= currentLogFile.length();
+        try {
+          currentLogOutputStream.close();
+          // if system down before delete, we can use this to delete file during recovery
+          maxAvailableTime = getFileTime(currentLogFile);
+          serializeMeta(meta);
+
+          currentLogFile.delete();
+          currentLogOutputStream = new FileOutputStream(getCurrentLogFile());
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+
+        logFileList.remove(logFileList.size() - 1);
+      }
+      // else we just truncate it
+      else {
+        try {
+          currentLogOutputStream.getChannel().truncate(getCurrentLogFile().length() - size);
+          break;
+        } catch (IOException e) {
+          logger.error("Error in log serialization: " + e.getMessage());
+        }
+      }
+    }
+
+  }
+
+  @Override
+  public void removeFirst(int num) {
+    firstLogPosition += num;
+    for (int i = 0; i < num; i++) {
+      removedLogSize += logSizeDeque.removeFirst();
+    }
+
+    // firstLogPosition changed
+    serializeMeta(meta);
+
+    // do actual deletion
+    if (removedLogSize > maxRemovedLogSize) {
+      openNewLogFile();
```

Comment:
That won't happen because in openNewLogFile, we firstly call actuallyDeleteFile, in this function we call adjustNextThreshold, maxRemovedLogSize will become the first log file length.

Reply:
Is it possible that the log file becomes very small?
For example:
When you close the 1st file, the length of the 1st file is 100, so you set maxRemovedLogSize to 100. But when removedLogSize reaches 100, the length of the 2nd file may only be 10, so you will set maxRemovedLogSize to 10.